File: src/main/java/org/jackhuang/compactwatermills/common/inventory/SlotInventorySlot.java
package org.jackhuang.compactwatermills.common.inventory;

import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;

public class SlotInventorySlot extends Slot {
	public final InventorySlot invSlot;
	public final int index;
	public final int stackLimit;
	
	public SlotInventorySlot(InventorySlot invSlot, int index,
			int xDisplayPosition, int yDisplayPosition) {
		this(invSlot, index, xDisplayPosition, yDisplayPosition, -9999);
	}

	public SlotInventorySlot(InventorySlot invSlot, int index,
			int xDisplayPosition, int yDisplayPosition, int stackLimit) {
		super(invSlot.base, -1, xDisplayPosition, yDisplayPosition);

		this.invSlot = invSlot;
		this.index = index;
		this.stackLimit = stackLimit;
	}

	public boolean isItemValid(ItemStack itemStack) {
		return this.invSlot.accepts(itemStack);
	}

	public ItemStack getStack() {
		return this.invSlot.get(this.index);
	}

	public void putStack(ItemStack itemStack) {
		this.invSlot.put(this.index, itemStack);
		onSlotChanged();
	}

	public ItemStack decrStackSize(int amount) {
		ItemStack itemStack = this.invSlot.get(this.index);
		if (itemStack == null)
			return null;

		if (itemStack.stackSize <= amount) {
			this.invSlot.put(this.index, null);
			onSlotChanged();

			return itemStack;
		}
		ItemStack ret = itemStack.copy();
		ret.stackSize = amount;

		itemStack.stackSize -= amount;
		onSlotChanged();

		return ret;
	}

	public boolean isSlotInInventory(IInventory inventory, int index) {
		if (inventory != this.invSlot.base)
			return false;

		for (InventorySlot invSlot : this.invSlot.base.invSlots) {
			if (index < invSlot.size()) {
				return index == this.index;
			}
			index -= invSlot.size();
		}

		return false;
	}
	
	@Override
	public int getSlotStackLimit() {
		if(stackLimit == -9999) return super.getSlotStackLimit();
		return stackLimit;
	}
}

File: src/main/kotlin/waterpower/common/block/inventory/SlotInventorySlot.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.block.inventory

import net.minecraft.inventory.IInventory
import net.minecraft.inventory.Slot
import net.minecraft.item.ItemStack
import waterpower.util.copyWithNewCount


class SlotInventorySlot(val invSlot: InventorySlot, val index: Int, xDisplayPosition: Int, yDisplayPosition: Int, val stackLimit: Int = -9999)
    : Slot(invSlot.tileEntity, -1, xDisplayPosition, yDisplayPosition) {

    override fun isItemValid(itemStack: ItemStack): Boolean {
        return this.invSlot.accepts(itemStack)
    }

    override fun getStack() = this.invSlot[this.index]

    override fun putStack(itemStack: ItemStack) {
        this.invSlot.put(this.index, itemStack)
        onSlotChanged()
    }

    override fun decrStackSize(amount: Int): ItemStack {
        val itemStack = this.invSlot[this.index]
        if (itemStack.isEmpty || amount <= 0) return ItemStack.EMPTY

        if (itemStack.count <= amount) {
            this.invSlot.put(this.index, ItemStack.EMPTY)
            onSlotChanged()

            return itemStack
        } else {
            val ret = itemStack.copyWithNewCount(amount)
            itemStack.shrink(amount)
            onSlotChanged()

            return ret
        }
    }

    fun isSlotInInventory(inventory: IInventory, index: Int): Boolean {
        var i = index
        if (inventory !== this.invSlot.tileEntity)
            return false

        for (s in this.invSlot.tileEntity.invSlots) {
            if (i < s.size()) {
                return i == this.index
            }
            i -= s.size()
        }

        return false
    }

    override fun getSlotStackLimit() =
            if (stackLimit == -9999)
                super.getSlotStackLimit()
            else stackLimit

}